Add timed callback dispatcher to security core

jrebelo opened the pull request:
Implement the timed callback dispatcher for the DDS Security core.

Signed-off-by: Joao Rebelo <jrebelo@s2e-systems.com>

dennis-adlink commented on src/security/core/src/dds_security_timed_cb.c:
```diff
@@ -0,0 +1,330 @@
+        ddsrt_threadattr_t attr;
+        ddsrt_threadattr_init(&attr);
+        osres = ddsrt_thread_create(g_thread_ptr, "security_dispatcher", &attr, timed_dispatcher_thread, NULL);
+        if (osres == DDS_RETCODE_ERROR)
```
The test for failure of thread creation should be `osres != DDS_RETCODE_OK` (e.g. the Windows abstration could return `DDS_RETCODE_OUT_OF_RESOURCES` because it allocates memory for the thread context)

jrebelo replied:
I have replaced the condition as suggested.

dennis-adlink commented on src/security/core/src/dds_security_timed_cb.c:
```diff
@@ -0,0 +1,330 @@
+        if (osres == DDS_RETCODE_ERROR)
+        {
+            fprintf(stderr, "Error: cannot create thread security_dispatcher");
+            exit(2);
```
Using the `DDS_FATAL` macro is the preferred way of handling a fatal error. This will result in a call to `dds_log()` which will call `abort()`

jrebelo replied:
I replaced the two lines by DDS_FATAL.

jrebelo commented:
The tests that occasionally fail on some of the platforms are not related to the changes introduced.

eboasson commented on src/security/core/src/dds_security_timed_cb.c:
```diff
@@ -0,0 +1,330 @@
+
+struct queue_event_t
+{
+    struct queue_event_t *next;
```
Implementing a priority queue with a linked list is "not so efficient". I'd much rather see it implemented using the priority queue in ddsrt (see ``fibheap.c``), just like in ``q_xevent.c`` or the lease management. It'd also eliminate most of the code, adding an event becomes a simple call to ``insert`` and checking/extracting the next event becomes ``ddsrt_fibheap_min``/``ddsrt_fibheap_extract_min``.

You would still have to keep track of the events associated with a dispatcher maintaining the current behaviour of ``dds_security_timed_dispatcher_free``, but you'd eliminate the scanning of the list of events.

jrebelo replied:
The reason why it is currently implemented as a linked list seems to be because multiple dispatchers with multiple event can simultaneously coexist on one thread. There are two options: 1. If we replace the list by the heap we either need to keep a dispatcher list with each its own fibheap and iterate through the list for finding which events have timedout. Keeping one heap and multiple dispatchers in the way it is currently done is not really feasible since nodes in the middle of the tree might need to be removed. 2. There is only one dispatcher per thread and all the resources are allocated to that dispatcher and duplicated if more dispatchers are needed.

Would you say it is better to have the linked list as it currently is or to keep a list of dispatcher with each dispatcher having its own fibheap?

eboasson replied:
I'm not sure I follow: right now there is a single priority queue of events (implemented as a sorted double-linked list), with the only unusual thing being that it skips any events associated with disabled dispatchers. (I don't know when/how/if/why that case is important, but I presume it is.)

Switching to a proper priority queue would mean never having events associated with a disabled dispatcher in the queue. Presumably one would let each disabled dispatcher have a list of events or a priority queue, to be added to the event queue when the dispatcher is enabled. Remains the question what to do when disabling a dispatcher. I'd just keep the current doubly-linked list to track events per dispatcher (minus the ordering).

That way the enable/disable operations get a bit more complicated because they need to add/remove the events from the priority queue, but the upside is that adding and processing become simpler.

jrebelo replied:
Thanks for the feedback @eboasson. Unfortunately I also can't clarify what is the use case for enabling/disabling dispatchers but I also assume it is important to keep it.

If I understand your idea correctly, we would keep the information about events in two places, the priority queue and the unordered linked list. If the dispatcher is disabled, events remain in the linked list but are removed from the queue. Once the dispatcher is enabled they are added again to the queue from the linked list. Is my understanding correct?

The main issue I have with that approach is that when an event is triggered it will have to be removed both from the priority queue and the linked list. I assume this to be the most common case and handling it will increase complexity. The second thing is that removing elements from the queue by disabling the dispatcher might result in non-minimum elements having to be removed which can be a costly operation for the queue.

I would see as a better alternative to this having a list of dispatchers, each with its own priority queue.

eboasson replied:
> Is my understanding correct?

Yes.

> The main issue I have with that approach is that when an event is triggered it will have to be removed both from the priority queue and the linked list. I assume this to be the most common case and handling it will increase complexity.

True; and I agree one would assume this to be a common case. I understand in principle how this contraption is/can be useful to the security implementation, but I don't know the details of how it will be used. That means any consideration about common cases versus uncommon case is a bit of a guess.

The model I proposed at least has the advantage that the only potentially expensive operations are enabling/disabling a dispatcher, and those seem to be unlikely to be common.

> The second thing is that removing elements from the queue by disabling the dispatcher might result in non-minimum elements having to be removed which can be a costly operation for the queue.

Removing an arbitrary element (in this implementation) means effectively a decrease-key + extract-min, so worst-case would be O(n log n) or thereabouts (all entries in the queue are owned by the dispatcher). That seems okayish.

> I would see as a better alternative to this having a list of dispatchers, each with its own priority queue.

That should work, too. Instead of a list you could also do a priority queue of active dispatchers, with each dispatcher scheduled based on the earliest event for that dispatcher.

Perhaps @MarcelJordense can shed some light on how this thing is intended to be used? It would be nice to have a better idea of that ...

MarcelJordense replied:
Both the authentication and the access_control plugin provide an interface to set a listener. Note that on each plugin a single listener can be set. The listener is used to inform the user of the authentication and access_control, which is the ddsi core, of identity status changes (not implemented) or of certificates that are invoked (expired). The set_listener is called when initializing the plugin and removed when the plugin is unloaded. Thus the timed callback functionality should allow both the authentication and the access_control plugin to set callbacks on the expiry time of certificates and remove the timeout callback when the corresponding certificate is removed when for example a local or remote participant is deleted.

stkimmer replied:
@MarcelJordense Thanks for the context clarification. Could you also clarify if the enabling/disabling of dispatcher is used or needed? The current interface allows creating/deleting dispatchers and enabling/disabling. The latter makes the implementation much harder.

MarcelJordense replied:
Yes enabling/disabling of dispatchers is needed. A dispatcher is associated with a listener that is set on either the authentication and access_control plugin. Thus a dispatcher is enabled when the plugin is initialized and disabled when the plugin is removed.

stkimmer commented on src/security/core/src/dds_security_timed_cb.c:
```diff
@@ -0,0 +1,330 @@
+};
+
+
+static ddsrt_mutex_t g_lock;
```
I would be better to use ddsrt_once indeed. However here is a counting of dispatchers included that allows automatic starting and stopping of the thread for created dispatchers. As such the user does not need to call any init/deinit. Could you tell what is broken with it?

eboasson replied:
What's broken is just the usual with such loops: calling init in parallel to fini you can end up with a schedule where init re-initializes everything while fini is still cleaning up; you need memory barriers to ensure the initialisation is actually globally visible before "initialized" can be read as true in line 101; "initialized" is used as-if atomic reads/writes are supported and as-if the compiler won't reorder it ...

If you tie it to the per-domain or to the global initialisation and teardown you avoid all those problems here. I'd do it in the per-domain ones and moreover put the global state in a dynamically allocated struct. That way the domains become wholly independent.

stkimmer commented:
We added a new/free pair for keeping an independent object instead of using global variables. The callback list is now implemented by using the fibheap priority queue for each dispatcher. To keep the functionality of being able to enable/disable specific dispatchers, dispatchers are kept in a linked list.  
Note that this changes the public interface and as such other pull request may need to be changed as well.

eboasson commented:
Thanks @jrebelo and @stkimmer. I remain thoroughly unconvinced by the interface (with the enabling and disabling in this way) despite @MarcelJordense’s comment regarding the use, but (pursuant to policy regarding this particular branch, &c.) I’ll reserve judgment until I have the complete picture. Hopefully it either will start making sense, and otherwise it is hoped that this decision will allow the implementation of DDS Security to proceed more efficiently than it otherwise would, if at the cost of additional rework.

However, I do have a number of review comments regarding various details in this particular bit of code:
* ``struct dds_security_timed_cb_data`` still has an ``init_cnt``, but it is no longer used.
* ``struct dds_security_timed_cb_data`` has a pointer to the thread id, but there always is a thread and so the indirection shouldn’t be there.
* One ought to check/modify ``tcb->terminate`` within the lock, but then obviously checking it inside the inner loop in ``timed_dispatcher_thread`` becomes of rather little value.
* I’m not convinced by the decision to hold the lock across the invocation of the event handler, but that may be acceptable given the use cases.
* ``dds_security_timed_dispatcher_free `` has a really overcomplicated loop for removing the dispatcher from the list: whether ``dispatched_node->next`` is a null pointer or makes no difference.

If it weren’t for the locking issue, I’d be willing to accept it as-is. It doesn’t make too much sense to keep at this little bit of code if it all has to be reworked anyway because of changes to the interface …

jrebelo commented:
Thank you for the feedback. We agree that there might be a better/simpler way to reach the goal here but, as you say, without a complete picture of the use cases it is indeed hard to rationalize about it. That being said, refactoring later will probably be a good idea anyway.

Regarding your comments:
- We removed the unused init_cnt from the struct
- We removed the indirection
- tcb->terminate is modified within the lock and the (now unnecessary) checks in the loop removed. One of the checks remained to handle the case when the thread is initialised and terminated in quick succession which can cause a deadlock.
- with the current architecture we don't know how to do different. Can be also part of a later refactoring depending on what changes in the architecture are done
- we simplified the loop



On 17 Jan 2020 15:03, eboasson <notifications@github.com> wrote:

Thanks @jrebelo<https://github.com/jrebelo> and @stkimmer<https://github.com/stkimmer>. I remain thoroughly unconvinced by the interface (with the enabling and disabling in this way) despite @MarcelJordense<https://github.com/MarcelJordense>’s comment regarding the use, but (pursuant to policy regarding this particular branch, &c.) I’ll reserve judgment until I have the complete picture. Hopefully it either will start making sense, and otherwise it is hoped that this decision will allow the implementation of DDS Security to proceed more efficiently than it otherwise would, if at the cost of additional rework.

However, I do have a number of review comments regarding various details in this particular bit of code:

  *   struct dds_security_timed_cb_data still has an init_cnt, but it is no longer used.
  *   struct dds_security_timed_cb_data has a pointer to the thread id, but there always is a thread and so the indirection shouldn’t be there.
  *   One ought to check/modify tcb->terminate within the lock, but then obviously checking it inside the inner loop in timed_dispatcher_thread becomes of rather little value.
  *   I’m not convinced by the decision to hold the lock across the invocation of the event handler, but that may be acceptable given the use cases.
  *   dds_security_timed_dispatcher_free has a really overcomplicated loop for removing the dispatcher from the list: whether dispatched_node->next is a null pointer or makes no difference.

If it weren’t for the locking issue, I’d be willing to accept it as-is. It doesn’t make too much sense to keep at this little bit of code if it all has to be reworked anyway because of changes to the interface …

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/eclipse-cyclonedds/cyclonedds/pull/340?email_source=notifications&email_token=AAU2RIZLLOY3RATMOR6ARWTQ6G3EJA5CNFSM4JRICZG2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJHYVCI#issuecomment-575638153>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAU2RI6REJUZSWLNWRW4QTDQ6G3EJANCNFSM4JRICZGQ>.

eboasson commented:
Closed via #374 after rebasing.
